src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -130,14 +134,85 @@ public static class TagHelperDescriptorFactory
                         assemblyName,
                         attributeDescriptors,
                         attribute,
+                        allowedChildren,
                         typeDesignTimeDescriptor));
         }
 
+        private static IEnumerable<string> GetAllowedChildren(TypeInfo typeInfo, ErrorSink errorSink)
+        {
+            var restrictChildrenAttribute = typeInfo.GetCustomAttribute<RestrictChildrenAttribute>(inherit: true);
+            if (restrictChildrenAttribute == null)
+            {
+                return null;
+            }
+
+            var allowedChildren = restrictChildrenAttribute.ChildTagNames;
+            var validAllowedChildren = GetValidAllowedChildren(allowedChildren, typeInfo.FullName, errorSink);
+
+            if (validAllowedChildren.Any())
+            {
+                return validAllowedChildren;
+            }
+            else
+            {
+                // All allowed children were invalid, return null to indicate that any child is acceptable.
+                return null;
+            }
+        }
+
+        // Internal for unit testing
+        internal static IEnumerable<string> GetValidAllowedChildren(
+            IEnumerable<string> allowedChildren,
+            string tagHelperName,
+            ErrorSink errorSink)
+        {
+            var validAllowedChildren = new List<string>();
+
+            foreach (var name in allowedChildren)
+            {
+                var valid = true;
+
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.FormatTagHelperDescriptorFactory_InvalidRestrictChildrenAttributeNameNullWhitespace(
+                            tagHelperName));
+
+                    valid = false;
+                }
+                else
+                {
+                    foreach (var character in name)
+                    {
+                        if (char.IsWhiteSpace(character) || InvalidNonWhitespaceNameCharacters.Contains(character))
+                        {
+                            errorSink.OnError(
+                                SourceLocation.Zero,
+                                Resources.FormatTagHelperDescriptorFactory_InvalidRestrictChildrenAttributeName(
+                                    name,
+                                    tagHelperName,
+                                    character));
+                            valid = false;
+                        }
+                    }
```

Comment:
Suggest refactoring slightly so that this bit is shared with `ValidateName()`. Also the resource should be similar to `FormatTargetElementAttribute_InvalidName`.